File: src/main/java/cz/encircled/eplayer/view/fx/components/AppMenuBar.java
package cz.encircled.eplayer.view.fx.components;

import cz.encircled.eplayer.core.ApplicationCore;
import cz.encircled.eplayer.model.MediaType;
import cz.encircled.eplayer.service.event.Event;
import cz.encircled.eplayer.util.Localization;
import cz.encircled.eplayer.view.fx.FxView;
import javafx.application.Platform;
import javafx.event.ActionEvent;
import javafx.event.EventHandler;
import javafx.scene.control.*;
import javafx.scene.input.KeyCode;
import javafx.scene.input.KeyCodeCombination;
import javafx.scene.input.KeyCombination;
import org.jetbrains.annotations.NotNull;
import uk.co.caprica.vlcj.player.TrackDescription;

import java.util.Iterator;
import java.util.List;
import java.util.stream.Collectors;

/**
 * @author Encircled on 27/09/2014.
 */
public class AppMenuBar {

    private FxView fxView;

    private ApplicationCore core;

    private Menu subtitles;

    private Menu audioTracks;

    public AppMenuBar(ApplicationCore core, FxView fxView) {
        this.core = core;
        this.fxView = fxView;
        init();
    }

    @NotNull
    public MenuBar getMenuBar() {
        MenuBar menuBar = new MenuBar();
        menuBar.getMenus().addAll(getFileMenu(), getViewMenu(), getMediaMenu(), getToolsMenu());
        return menuBar;
    }

    @NotNull
    private Menu getFileMenu() {
        Menu file = new Menu(Localization.file.ln());

        MenuItem open = new MenuItem(Localization.open.ln());
        open.setAccelerator(new KeyCodeCombination(KeyCode.O, KeyCombination.CONTROL_DOWN));
        open.setOnAction(event -> fxView.openMedia());

        MenuItem exit = new MenuItem(Localization.exit.ln());
        exit.setAccelerator(new KeyCodeCombination(KeyCode.Q, KeyCombination.CONTROL_DOWN));
        exit.setOnAction(event -> core.exit());

        file.getItems().addAll(open, exit);
        return file;
    }

    @NotNull
    private Menu getViewMenu() {
        Menu view = new Menu(Localization.view.ln());

        CheckMenuItem fullScreen = new CheckMenuItem(Localization.fullScreen.ln());
        fxView.getPrimaryStage().fullScreenProperty().addListener((observable, oldValue, newValue) -> fullScreen.setSelected(newValue));
        fullScreen.setOnAction(event -> fxView.toggleFullScreen());
        fullScreen.setAccelerator(new KeyCodeCombination(KeyCode.F, KeyCombination.CONTROL_DOWN));
        fullScreen.setSelected(fxView.isFullScreen());

        CheckMenuItem fitScreen = new CheckMenuItem(Localization.fitScreen.ln());
        fxView.getPlayerScreen().fitToScreenProperty().addListener((observable, oldValue, newValue) -> fitScreen.setSelected(newValue));
        fitScreen.setOnAction(event -> fxView.getPlayerScreen().toggleFitToScreen());
        fitScreen.setAccelerator(new KeyCodeCombination(KeyCode.F, KeyCombination.CONTROL_DOWN, KeyCodeCombination.SHIFT_DOWN));
        fitScreen.setSelected(fxView.getPlayerScreen().fitToScreenProperty().get());

        view.getItems().addAll(fullScreen, fitScreen);
        return view;
    }

    @NotNull
    private Menu getMediaMenu() {
        Menu media = new Menu(Localization.media.ln());

        MenuItem play = new MenuItem(Localization.play.ln());
        play.setOnAction(event -> core.getMediaService().toggle());
        core.getEventObserver().listenFxThread(Event.playingChanged, (event, isPlaying, arg2) -> {
            play.setText(isPlaying ? Localization.pause.ln() : Localization.play.ln());
        });

        subtitles = new Menu(Localization.subtitles.ln());
        audioTracks = new Menu(Localization.audioTrack.ln());
        subtitles.setDisable(true);
        audioTracks.setDisable(true);

        media.getItems().addAll(play, subtitles, audioTracks);

        return media;
    }

    @NotNull
    private Menu getToolsMenu() {
        Menu tools = new Menu(Localization.tools.ln());

        MenuItem openQn = new MenuItem(Localization.openQuickNavi.ln());
        openQn.setAccelerator(new KeyCodeCombination(KeyCode.N, KeyCombination.CONTROL_DOWN));
        openQn.setOnAction(event -> core.openQuickNavi());

        MenuItem deleteMissing = new MenuItem(Localization.deleteMissing.ln());
        deleteMissing.setOnAction(event -> new Thread(() -> {
            Iterator<MediaType> iterator = core.getCacheService().getCache().iterator();
            while (iterator.hasNext()) {
                MediaType m = iterator.next();
                if (!m.exists())
                    iterator.remove();
            }
            core.getCacheService().save();
            Platform.runLater(fxView.getQuickNaviScreen()::refreshCurrentTab);
        }).start());

        tools.getItems().addAll(openQn, deleteMissing);

        return tools;
    }

    public void init() {
        core.getEventObserver().listenFxThread(Event.subtitlesUpdated, (event, subTracks, arg2) -> {
            EventHandler<ActionEvent> subtitlesHandler = event1 -> {
                core.getMediaService().setSubtitles((int) ((RadioMenuItem) event1.getSource()).getUserData());
            };
            updateTrackMenu(subtitles, subTracks, core.getMediaService().getSubtitles(), subtitlesHandler);
        });

        core.getEventObserver().listenFxThread(Event.audioTracksUpdated, (event, subTracks, arg2) -> {

            EventHandler<ActionEvent> audioTrackHandler = event1 -> {
                core.getMediaService().setAudioTrack((int) ((RadioMenuItem) event1.getSource()).getUserData());
            };
            updateTrackMenu(audioTracks, subTracks, core.getMediaService().getAudioTrack(), audioTrackHandler);

        });

        fxView.screenChangeProperty().addListener((observable, oldValue, newValue) -> {
            if (FxView.QUICK_NAVI_SCREEN.equals(newValue)) {
                audioTracks.setDisable(true);
                subtitles.setDisable(true);
            }
        });
    }

    private void updateTrackMenu(@NotNull Menu menu, @NotNull List<TrackDescription> trackDescriptions, final int selected, EventHandler<ActionEvent> eventHandler) {
        ToggleGroup toggleGroup = new ToggleGroup();
        menu.getItems().setAll(trackDescriptions.stream().map(s -> {
            RadioMenuItem menuItem = new RadioMenuItem(s.description());

            menuItem.setSelected(s.id() == selected);
            menuItem.setToggleGroup(toggleGroup);
            menuItem.setUserData(s.id());
            menuItem.setOnAction(eventHandler);

            return menuItem;
        }).collect(Collectors.toList()));
        menu.setDisable(trackDescriptions.isEmpty());
    }

}

File: src/main/java/cz/encircled/eplayer/view/fx/components/AppMenuBar.kt
package cz.encircled.eplayer.view.fx.components

import cz.encircled.eplayer.core.ApplicationCore
import cz.encircled.eplayer.model.GenericTrackDescription
import cz.encircled.eplayer.service.event.Event
import cz.encircled.eplayer.util.Localization
import cz.encircled.eplayer.view.fx.FxView
import cz.encircled.eplayer.view.fx.UiDataModel
import cz.encircled.eplayer.view.fx.addNewValueListener
import javafx.event.ActionEvent
import javafx.event.EventHandler
import javafx.scene.control.*
import javafx.scene.input.KeyCode
import javafx.scene.input.KeyCodeCombination
import javafx.scene.input.KeyCombination

/**
 *
 * @author Encircled on 27/09/2014.
 */
class AppMenuBar(private val core: ApplicationCore, private val fxView: FxView, private val dataModel: UiDataModel) {
    lateinit var subtitles: Menu
    lateinit var audioTracks: Menu

    fun getMenuBar(): MenuBar = MenuBar(fileMenu(), viewMenu(), mediaMenu(), toolsMenu())

    private fun fileMenu(): Menu {
        val file = Menu(Localization.file.ln())

        val open = MenuItem(Localization.open.ln())
        open.accelerator = KeyCodeCombination(KeyCode.O, KeyCombination.CONTROL_DOWN)
        open.onAction = EventHandler { fxView.openMediaChooser() }

        val exit = MenuItem(Localization.exit.ln())
        exit.accelerator = KeyCodeCombination(KeyCode.Q, KeyCombination.CONTROL_DOWN)
        exit.onAction = EventHandler { core.exit() }

        file.items.addAll(open, exit)
        return file
    }

    private fun viewMenu(): Menu {
        val view = Menu(Localization.view.ln())

        val fullScreen = CheckMenuItem(Localization.fullScreen.ln())
        fxView.primaryStage.fullScreenProperty().addNewValueListener { fullScreen.isSelected = it }
        fullScreen.onAction = EventHandler { fxView.toggleFullScreen() }
        fullScreen.accelerator = KeyCodeCombination(KeyCode.F, KeyCombination.CONTROL_DOWN)
        fullScreen.isSelected = fxView.isFullScreen

        val fitScreen = CheckMenuItem(Localization.fitScreen.ln())
        dataModel.fitToScreen.addNewValueListener { fitScreen.isSelected = it }
        fitScreen.onAction = EventHandler { dataModel.toggleFitToScreen() }
        fitScreen.accelerator = KeyCodeCombination(KeyCode.F, KeyCombination.CONTROL_DOWN, KeyCodeCombination.SHIFT_DOWN)
        fitScreen.isSelected = dataModel.fitToScreen.get()

        val back = MenuItem(Localization.back.ln())
        back.onAction = EventHandler { core.back() }
        back.accelerator = KeyCodeCombination(KeyCode.ESCAPE)
        view.items.addAll(fullScreen, fitScreen)

        return view
    }

    private fun mediaMenu(): Menu {
        val media = Menu(Localization.media.ln())

        val play = MenuItem(Localization.play.ln())
        play.onAction = EventHandler { core.mediaService.toggle() }
        play.accelerator = KeyCodeCombination(KeyCode.SPACE)
        Event.playingChanged.listenFxThread { isPlaying: Boolean ->
            play.text = if (isPlaying) Localization.pause.ln() else Localization.play.ln()
        }

        subtitles = Menu(Localization.subtitles.ln())
        audioTracks = Menu(Localization.audioTrack.ln())
        subtitles.isDisable = true
        audioTracks.isDisable = true

        media.items.addAll(play, subtitles, audioTracks)
        return media
    }

    private fun toolsMenu(): Menu {
        val tools = Menu(Localization.tools.ln())

        val openQn = MenuItem(Localization.openQuickNavi.ln())
        openQn.accelerator = KeyCodeCombination(KeyCode.N, KeyCombination.CONTROL_DOWN)
        openQn.onAction = EventHandler { core.openQuickNavi() }

        val deleteMissing = MenuItem(Localization.deleteMissing.ln())
        deleteMissing.onAction = EventHandler {
            Thread {
                core.cacheService.getCached().filter {
                    !it.mediaFile().exists()
                }.forEach {
                    core.cacheService.deleteEntry(it)
                }
                core.cacheService.save()
                fxView.quickNaviController.forceRefresh()
            }.start()
        }

        tools.items.addAll(openQn, deleteMissing)
        return tools
    }

    init {
        Event.subtitlesUpdated.listenFxThread { tracks: List<GenericTrackDescription> ->
            updateTrackMenu(subtitles, tracks, core.mediaService.subtitles) {
                core.mediaService.subtitles = (it.source as RadioMenuItem).userData as Int
            }
        }

        Event.audioTracksUpdated.listenFxThread { tracks: List<GenericTrackDescription> ->
            updateTrackMenu(audioTracks, tracks, core.mediaService.audioTrack) {
                core.mediaService.audioTrack = (it.source as RadioMenuItem).userData as Int
            }
        }

        fxView.sceneChangeProperty.addNewValueListener {
            if (FxView.QUICK_NAVI_SCREEN == it) {
                audioTracks.isDisable = true
                subtitles.isDisable = true
            }
        }
    }

    private fun updateTrackMenu(menu: Menu, trackDescriptions: List<GenericTrackDescription>, selected: Int, eventHandler: EventHandler<ActionEvent>) {
        val toggleGroup = ToggleGroup()
        menu.items.setAll(trackDescriptions.map {
            RadioMenuItem(it.description).apply {
                isSelected = it.id == selected
                this.toggleGroup = toggleGroup
                userData = it.id
                onAction = eventHandler
            }
        })
        menu.isDisable = trackDescriptions.isEmpty()
    }

}